substratevm/src/com.oracle.svm.reflect/src/com/oracle/svm/reflect/serialize/hosted/SerializationFeature.java

```diff
@@ -96,8 +117,150 @@ public void duringSetup(DuringSetupAccess a) {
                         ConfigurationFile.SERIALIZATION.getFileName());
     }
 
+    private static GraphBuilderConfiguration buildLambdaParserConfig() {
+        GraphBuilderConfiguration.Plugins plugins = new GraphBuilderConfiguration.Plugins(new InvocationPlugins());
+        plugins.setClassInitializationPlugin(new NoClassInitializationPlugin());
+        return GraphBuilderConfiguration.getDefault(plugins).withEagerResolving(true);
+    }
+
+    @SuppressWarnings("try")
+    private static StructuredGraph createMethodGraph(ResolvedJavaMethod method, DebugContext debug) {
+        StructuredGraph graph = new StructuredGraph.Builder(debug.getOptions(), debug).method(method).build();
+        try (DebugContext.Scope ignored = debug.scope("ParsingToMaterializeLambdas")) {
+            GraphBuilderPhase lambdaParserPhase = new GraphBuilderPhase(buildLambdaParserConfig());
+            HighTierContext context = new HighTierContext(GraalAccess.getOriginalProviders(), null, OptimisticOptimizations.NONE);
+            lambdaParserPhase.apply(graph, context);
+        } catch (Throwable e) {
+            throw debug.handle(e);
+        }
+        return graph;
+    }
+
+    private static Class<?> getLambdaClassFromMemberField(Class<?> clazz, Object constantObject, String fieldName) {
+        try {
+            Field memberField = clazz.getDeclaredField(fieldName);
+            memberField.setAccessible(true);
+            Object object = memberField.get(constantObject);
+            Method getDeclaringClass = object.getClass().getDeclaredMethod("getDeclaringClass");
+            getDeclaringClass.setAccessible(true);
+            return (Class<?>) getDeclaringClass.invoke(object);
+        } catch (NoSuchFieldException | IllegalAccessException | NoSuchMethodException | InvocationTargetException exception) {
+            return null;
+        }
+    }
+
+    private static Class<?> getLambdaClassFromConstantObject(Class<?> clazz, Object constantObject) {
+        try {
+            Field instanceClassField = clazz.getDeclaredField("staticBase");
+            instanceClassField.setAccessible(true);
+            return (Class<?>) instanceClassField.get(constantObject);
+        } catch (IllegalAccessException | NoSuchFieldException ignored) {
+            return null;
+        }
+    }
+
+    /**
+     * It is not possible to avoid reflection when trying to get lambda class from
+     * {@link ConstantNode} instance. Lambda class is wrapped in the {@code ConstantNode#value}
+     * which is of type {@code DirectHotSpotObjectConstantImpl}. That class is protected and it is
+     * not visible in the {@link SerializationFeature}. The only way to access fields and methods of
+     * this class is through the reflection. There are 2 different ways of getting lambda from the
+     * {@code ConstantNode#value}:
```

Comment:
You need to go through `SnippetReflectionProvider.asObject`. You get the provider via `GraalAccess.getOriginalProviders` that you already call elsewhere.

Reply:
`SnippetReflectionProvider.asObject` used as suggested.